### PR TITLE
fix: support `.d.mts` and `.d.cts`

### PR DESCRIPTION
**Pull request by antfu**
https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext

For example, package `shikiji` is an ESM-only package that only ships `.mjs` and `.d.mts`. The ATA failed to work as failed to find `.d.ts` and start looking for `@types/shikiji`, which do not exist either:
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgcwKYwBLGQCwDbY4ypRwC+cAZlBCHAOQDOOwA1sAFbD0BQPAxhAB2jeAEM4AXjhiA7mODw0mQgVzEoACgCUQA

### CODE REVIEW

File: packages/sandbox/src/index.ts [hunk 1]
```diff
@@ -314,12 +314,15 @@ export const createTypeScriptSandbox = (
     return (firstJS && firstJS.text) || ""
   }
 
+  const isDtsFile = (name: string) => /\.d\.[cm]?ts$/.test(name)
```
Comment (jakebailey) [inline packages/sandbox/src/index.ts]: You may also want to support `.d.*.ts`, e.g. `.d.css.ts`. https://github.com/microsoft/DefinitelyTyped-tools/blob/7a8a32513cbbe5a2af5c3e39e056117b3c14c390/packages/utils/src/miscellany.ts#L97

Comment (antfu) [reply packages/sandbox/src/index.ts]: Interesting, good to know. Updated

**Review by orta — APPROVED** [review]

**Review by jakebailey — APPROVED** [review]
Kinda frustrating that there's no common way to access this one regex, but, oh well.

Not sure how much it matters but it may technically be more performant to store the regexp object off in a var somewhere and reuse it. not sure if V8 performs that optimization or not, though.

Also will require a package release.

File: packages/ts-twoslasher/src/index.ts [hunk 2]
```diff
@@ -742,7 +742,7 @@
     // Get the file which created the file we want to show:
     const emitFilename = handbookOptions.showEmittedFile || defaultFileName
     const emitSourceFilename =
-      fsRoot + emitFilename.replace(".jsx", "").replace(".js", "").replace(".d.ts", "").replace(".map", "")
+      fsRoot + emitFilename.replace(".jsx", "").replace(".js", "").replace(/\.d\.(.+\.)?[cm]?ts$/i, "").replace(".map", "")
```
[the remark this answers is not in the document]
Comment (jakebailey) [reply packages/ts-twoslasher/src/index.ts]: Ah, I suspect that this will need to be fixed.

[the remark this answers is not in the document]
Comment (antfu) [reply packages/ts-twoslasher/src/index.ts]: I am not so sure how to fix it, pushed a commit to try and it seems requires approval to run the CI

[the remark this answers is not in the document]
Comment (jakebailey) [reply packages/ts-twoslasher/src/index.ts]: I believe the change you made should do it, but we'll see if CodeQL is happy.

**Pull request merged**